File: offer/simple/PHP/twoSum.php
<?php

/**
 * 剑指 Offer 57. 和为s的两个数字
 * @Author: pengwei
 * @Date:   2021-05-23 18:27:40
 * @Last Modified by:   pengwei
 */

function twoSum($nums, $target) {
    // 方法1-hash
    // $map = [];
    // $len = count($nums);
    // for ($i = 0; $i < $len; $i++) {
    //     $map[$nums[$i]] = $nums[$i];
    // }
    // for ($i = 0; $i < $len; $i++) {
    //     $tmp = $target - $nums[$i];
    //     if (isset($map[$tmp])) {
    //         return [$nums[$i], $tmp];
    //     }
    // }
    // return [];

    // 方法2-双指针
    $i = 0;
    $len = count($nums) - 1;
    while ($i < $len) {
        if ($nums[$i] + $nums[$len] == $target) {
            return [$nums[$i], $nums[$len]];
        } elseif ($nums[$i] + $nums[$len] < $target) {
            ++$i;
        } else {
            --$len;
        }
    }
    return [];
}

File: offer/simple/Golang/twoSum.go
package main

import "fmt"

// 两数之和
// 剑指 Offer 57. 和为s的两个数字

func main() {

	nums := []int{1, 2, 3, 4}
	target := 6

	res := towSum(nums, target)
	fmt.Println(res)
}

func towSum(nums []int, target int) []int {

	// 方法1
	//l := len(nums)
	//tmp := 0
	//m := make(map[int]int, l)
	//s := make([]int, 0)
	//for _, v := range nums {
	//	tmp = target - v
	//	if val, ok := m[tmp]; ok {
	//		s = append(s, val, v)
	//		break
	//	}
	//	m[v] = v
	//}
	//return s

	// 方法2
	numsLen := len(nums)
	if numsLen < 1 {
		return nums
	}

	i := 0
	j := numsLen - 1
	result := make([]int, 0)
	for i < j {
		if nums[i] + nums[j] == target {
			result = append(result, nums[i], nums[j])
			break
		} else if nums[i] + nums[j] > target {
			j--
		} else {
			i++
		}
	}
	return result
}
